packages/wonder-blocks-form/src/components/text-field.md

```diff
@@ -54,31 +54,215 @@ class TextFieldExample extends React.Component {
         super(props);
         this.state = {
             value: "Password123",
+            errorMessage: null,
         };
+        this.handleOnChange = this.handleOnChange.bind(this);
+        this.handleValidation = this.handleValidation.bind(this);
+        this.handleOnError = this.handleOnError.bind(this);
+    }
+
+    handleOnChange(newValue) {
+        this.setState({value: newValue});
+    }
+
+    handleValidation(value) {
+        if (value.length <= 8) {
+            return "Password must be at least 8 characters long";
+        }
+        if (!/\d/.test(value)) {
+            return "Password must contain a numeric value";
+        }
+        this.setState({errorMessage: null})
+    }
+
+    handleOnError(errorMessage) {
+        this.setState({errorMessage: errorMessage});
     }
 
     render() {
         return (
-            <TextField
-                id="tf-1"
-                type="password"
-                value={this.state.value}
-                onChange={(newValue) => this.setState({value: newValue})}
-            />
+            <View>
+                <TextField
+                    id="tf-1"
+                    type="password"
+                    value={this.state.value}
+                    validation={this.handleValidation}
+                    onChange={this.handleOnChange}
+                    onError={this.handleOnError}
+                />
+                {this.state.errorMessage && (
+                    <View>
+                        <Strut size={Spacing.xSmall_8} />
+                        <Text style={styles.errorMessage}>{this.state.errorMessage}</Text>
+                    </View>
+                )}
+            </View>
+        );
+    }
+}
+
+const styles = StyleSheet.create({
+    errorMessage: {
+        color: Color.red,
+        paddingLeft: Spacing.xxxSmall_4,
+    },
+});
+
+<TextFieldExample />
+```
+
+Email (with Validation)
+
+```js
+import {StyleSheet} from "aphrodite";
+import {View, Text} from "@khanacademy/wonder-blocks-core";
+import Color from "@khanacademy/wonder-blocks-color";
+import {Strut} from "@khanacademy/wonder-blocks-layout";
+import Spacing from "@khanacademy/wonder-blocks-spacing";
+import {TextField} from "@khanacademy/wonder-blocks-form";
+
+class TextFieldExample extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            value: "khan@khanacademy.org",
+            errorMessage: null,
+        };
+        this.handleOnChange = this.handleOnChange.bind(this);
+        this.handleValidation = this.handleValidation.bind(this);
+        this.handleOnError = this.handleOnError.bind(this);
+    }
+
+    handleOnChange(newValue) {
+        this.setState({value: newValue});
+    }
+
+    handleValidation(value) {
+        const emailRegex = /^[^@\s]+@[^@\s.]+\.[^@.\s]+$/;
+        if (!emailRegex.test(value)) {
+            return "Please enter a valid email";
+        }
+        this.setState({errorMessage: null})
```

Comment:
mmm I'm thinking that this function should only take care of validating the value without having to modify the state (or anything related about React lifecycle and state). I understand that you had to set state here to reset the error state in case the field is valid, but I wonder if we need to modify the `onError` prop to be called no matter if it's an error or not.

That said, I can see different approaches:
1. Modify `onError` to be called every time the validation is performed.
2. Rename `onError` to something more meaningful: e.g. `onValidation(errorMessage: string)`.

IMO using `onValidation` feels more natural in the sense that it's the function that is called after performing `validation`, but ofc I'd like to hear your thoughts on this, and probably explore other possibilities.

Reply:
BTW Thanks for adding this example! It helped to identify this situation ahead of time :)

Reply:
I see.... I agree with the `onValidation` route.

At first, I viewed it as redundant since `validation` will call `onValidation`, so why not just put everything in the `validation` function? 

But then thinking deeper about it, I can see why it would be better to have both:
* `validation` will **only** handle validating the value
* `onValidation` will handle or modify any states or variables.